src/aleph/web/controllers/storage.py

```diff
@@ -114,59 +117,168 @@ class StorageMetadata(pydantic.BaseModel):
     sync: bool
 
 
-class UploadedFile(Protocol):
+class UploadedFile:
+    def __enter__(self):
+        raise NotImplementedError
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.cleanup()
+
+    def cleanup(self):
+        pass
+
+    async def read_and_validate(self):
+        pass
+
     @property
     def size(self) -> int:
-        ...
+        raise NotImplementedError
 
     @property
-    def content(self) -> Union[str, bytes]:
-        ...
+    def content(self) -> bytes:
+        raise NotImplementedError
 
+    @property
+    def file(self):
+        raise NotImplementedError
+
+    def get_hash(self) -> str:
+        raise NotImplementedError
 
-class MultipartUploadedFile:
-    _content: Optional[bytes]
-    size: int
 
-    def __init__(self, file_field: FileField, size: int):
+class MultipartUploadedFile(UploadedFile):
+    def __init__(self, file_field: BodyPartReader, max_size: int):
         self.file_field = file_field
-        self.size = size
-        self._content = None
+        self.max_size = max_size
+        self._hash = None
+        try:
+            self._temp_file = tempfile.NamedTemporaryFile(delete=False)
```

Comment:
I think it would make more sens to create the temporary file in `__enter__` to only create it when you need it and having hanging temporary files before the workflow has effectively started.